### PR TITLE
Feat/social share key

### PR DESCRIPTION
**Pull request by roman-khomenko-dev**
Implement social sharing key feature

### CODE REVIEW

File: lib/chat/key_share.ex [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+  alias Chat.Identity
+  alias Chat.{Actor, ChunkedFiles, ChunkedFilesMultisecret}
+  alias Chat.Upload.UploadKey
+  alias Chat.{Dialogs, Dialogs.Dialog, Messages}
+
+  alias ChatWeb.MainLive.Page
+
+  def generate_key_shares({me, rooms, users}) do
+    share_count = Enum.count(users)
+    base_key = Actor.new(me, rooms, %{}) |> Actor.to_encrypted_json("") |> Base.encode64()
```
Comment (sergey-lukianov) [inline lib/chat/key_share.ex:13]: We are sharing user identity only. Rooms would be in his `My notes`

File: lib/chat/key_share.ex [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+           file_secret <- ChunkedFiles.new_upload(file_key) do
+        save({file_key, share.key}, {file_info.size, file_secret})
+        send(entry, dialog, me, {file_key, file_secret, time})
+        File.rm!(file_path)
+      end
+    end)
+  end
+
+  defp file_path(%Identity{name: name} = me),
+    do: "/tmp/#{name}-#{Enigma.short_hash(me)}.social_part"
```
Comment (sergey-lukianov) [inline lib/chat/key_share.ex:47]: I doubt it is `short_hash(identity)` it is likely to be `short_hash(pub_key)`

File: lib/chat/key_share.ex [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+      pub_key: Base.encode16(b_key, case: :lower),
+      type: :dialog
+    }
+  end
+
+  defp entry(file_info, {me, path}) do
+    %{
+      client_last_modified:
+        file_info.mtime |> Timex.format("{YYYY}-{M}-{D} {h24}:{m}:{s}") |> elem(1),
+      client_name: "#{me.name}-#{Enigma.short_hash(me)}.social_part",
```
Comment (sergey-lukianov) [inline lib/chat/key_share.ex:66]: use the function here

File: lib/chat/key_share.ex [hunk 4]
```diff
@@ -0,0 +1,88 @@
+defmodule Chat.KeyShare do
+  @moduledoc "Manipulate social sharing keys"
+
+  alias Chat.Identity
+  alias Chat.{Actor, ChunkedFiles, ChunkedFilesMultisecret}
+  alias Chat.Upload.UploadKey
+  alias Chat.{Dialogs, Dialogs.Dialog, Messages}
+
+  alias ChatWeb.MainLive.Page
```
Comment (sergey-lukianov) [inline lib/chat/key_share.ex:9]: `Chat` should not use `ChatWeb`

File: lib/chat/key_share.ex [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+    base_key = Actor.new(me, rooms, %{}) |> Actor.to_encrypted_json("") |> Base.encode64()
+    len_part = ceil(String.length(base_key) / share_count)
+
+    for i <- 0..(share_count - 1), into: [] do
+      start_idx = i * len_part
+      end_idx = start_idx + len_part - 1
+
+      %{
+        user: Enum.at(users, i),
+        key: String.slice(base_key, start_idx..end_idx)
```
Comment (sergey-lukianov) [inline lib/chat/key_share.ex:22]: use `Enigma.`

File: lib/chat/key_share.ex [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+    end)
+  end
+
+  defp file_path(%Identity{name: name} = me),
+    do: "/tmp/#{name}-#{Enigma.short_hash(me)}.social_part"
+
+  defp generate_file(path, key) do
+    File.write!(path, key)
+    File.stat!(path)
+  end
```
Comment (sergey-lukianov) [inline lib/chat/key_share.ex:52]: why we need temp file

**Review by sergey-lukianov — CHANGES_REQUESTED** [review]

File: lib/chat/key_share.ex (hunk 6 above)
Comment (roman-khomenko-dev) [reply lib/chat/key_share.ex:52]: temp file removed

File: lib/chat/key_share.ex (hunk 5 above)
Comment (roman-khomenko-dev) [reply lib/chat/key_share.ex:22]: done with Enigma.encrypt/3

File: lib/chat/key_share.ex (hunk 4 above)
Comment (roman-khomenko-dev) [reply lib/chat/key_share.ex]: changed to use Phoenix.PubSub for broadcasting

File: lib/chat/key_share.ex (hunk 2 above)
Comment (roman-khomenko-dev) [reply lib/chat/key_share.ex]: short_hash work same as for pub_key coz Identity implement Enigma.Hash.Protocol but i change it to use pub_key

File: lib/chat/key_share.ex (hunk 1 above)
Comment (roman-khomenko-dev) [reply lib/chat/key_share.ex]: rooms removed from key

File: lib/chat/key_share.ex [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+
+  defp broadcast(message, dialog) do
+    PubSub.broadcast!(
+      Chat.PubSub,
+      dialog
+      |> Dialogs.key()
+      |> then(&"dialog:#{&1}"),
+      {:dialog, {:new_dialog_message, message}}
+    )
```
Comment (sergey-lukianov) [inline lib/chat/key_share.ex:88]: I strive to keep `Chat` pure and `ChatWeb` to do most side effects

In this case I would have a function that generates parts for each user in Chat, and sending and broadcasting in ChatWeb

File: lib/chat/key_share.ex [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+           destination <- destination(dialog),
+           file_key <- UploadKey.new(destination, dialog.b_key, entry),
+           file_secret <- ChunkedFiles.new_upload(file_key) do
+        save({file_key, share.key}, {file_info.size, file_secret})
+        send(entry, dialog, me, {file_key, file_secret, file_info.time})
+      end
+    end)
+  end
+
+  def schema, do: %{users: {:array, :string}}
```
Comment (sergey-lukianov) [inline lib/chat/key_share.ex:46]: what is this for ?

Comment (roman-khomenko-dev) [reply lib/chat/key_share.ex]: it moved to KeyShareForm, use such schema to validate share users form, removed here

File: lib/chat/key_share.ex (hunk 7 above)
Comment (roman-khomenko-dev) [reply lib/chat/key_share.ex]: change KeyShare for only generate and save shares, sending process moved to live component

File: lib/chat/key_share.ex [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+  alias Chat.Upload.UploadKey
+
+  @threshold 4
+
+  def generate_key_shares({%Identity{private_key: private_key} = _me, users}) do
+    shares =
+      private_key
+      |> Enigma.hide_secret_in_shares(users |> Enum.count(), @threshold)
+
+    share_count = Enum.count(shares)
```
Comment (sergey-lukianov) [inline lib/chat/key_share.ex:15]: User count, then shares count ... they should be the same

File: lib/chat/key_share.ex [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+
+    for i <- 0..(share_count - 1), into: [] do
+      %{
+        user: Enum.at(users, i),
+        key: Enum.at(shares, i) |> Base.encode64()
+      }
+    end
```
Comment (sergey-lukianov) [inline lib/chat/key_share.ex:22]: almost Enum.zip

File: lib/chat/key_share.ex [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+  end
+
+  def save_shares(shares, {me, time_offset}) do
+    shares
+    |> Enum.reduce([], fn share, acc ->
```
Comment (sergey-lukianov) [inline lib/chat/key_share.ex:27]: It is more `Enum.map` than `Enum.reduce`

File: lib/chat_web/live/main_live/index.html.heex [hunk 12]
```diff
@@ -117,6 +117,15 @@
               <.icon id="upload" class="w-4 h-4 ml-2 fill-white" />
             </div>
           </button>
+          <button
+            class="mt-5 w-full h-12 border rounded-lg border-black/10 flex items-center justify-center t-download-key"
+            phx-click="logout-go-share"
+          >
+            <div class="flex items-center justify-between">
+              <span>Share the Keys</span>
```
Comment (sergey-lukianov) [inline lib/chat_web/live/main_live/index.html.heex:125]: Share the Key

File: lib/chat_web/live/main_live/page/key_share_form.ex [hunk 13]
```diff
@@ -0,0 +1,157 @@
...
+          class: "mt-5 w-full h-12 border-0 rounded-lg bg-grayscale text-white disabled:opacity-50",
+          disabled: !@changeset.valid?
+        ) %>
+      </.form>
+    </div>
+    """
+  end
+
+  defp selected_user(%{assigns: %{share_users: users}} = socket, user) do
+    case user in users do
```
Comment (sergey-lukianov) [inline lib/chat_web/live/main_live/page/key_share_form.ex:128]: consider MapSet for sets operations

File: lib/chat_web/live/main_live/page/key_share_form.ex [hunk 14]
```diff
@@ -0,0 +1,157 @@
...
+  defp send_share(%{
+         entry: entry,
+         dialog: dialog,
+         me: me,
+         file_info: {file_key, file_secret, time}
+       }) do
+    entry
+    |> Messages.File.new(file_key, file_secret, time)
+    |> Dialogs.add_new_message(me, dialog)
+    |> broadcast(dialog)
```
Comment (sergey-lukianov) [inline lib/chat_web/live/main_live/page/key_share_form.ex:143]: where is file_key await?

**Review by sergey-lukianov — CHANGES_REQUESTED** [review]

File: lib/chat/key_share.ex (hunk 9 above)
Comment (roman-khomenko-dev) [reply lib/chat/key_share.ex]: change to use amount = Enum.count(users)

File: lib/chat/key_share.ex (hunk 10 above)
Comment (roman-khomenko-dev) [reply lib/chat/key_share.ex]: done using Enum.zip_reduce/4 instead of for

File: lib/chat_web/live/main_live/page/key_share_form.ex (hunk 13 above)
Comment (roman-khomenko-dev) [reply lib/chat_web/live/main_live/page/key_share_form.ex:103]: :share_users assign changed to MapSet

File: lib/chat_web/live/main_live/page/key_share_form.ex (hunk 14 above)
Comment (roman-khomenko-dev) [reply lib/chat_web/live/main_live/page/key_share_form.ex:119]: add |> Dialogs.await_saved/2

**Review by sergey-lukianov — APPROVED** [review]

**Pull request merged**